test/Microsoft.AspNetCore.Authentication.Test/FacebookTests.cs

```diff
@@ -562,6 +556,66 @@ private void ConfigureDefaults(FacebookOptions o)
             Assert.Contains("custom=test", query);
         }
 
+        [Fact]
+        public async Task ChallengeWillIncludeScopeAsConfigured()
+        {
+            var server = CreateServer(
+                app => app.UseAuthentication(),
+                services =>
+                {
+                    services.AddAuthentication().AddFacebook(o =>
+                    {
+                        o.AppId = "Test App Id";
+                        o.AppSecret = "Test App Secret";
+                        o.Scope.Clear();
+                        o.Scope.Add("foo");
+                        o.Scope.Add("bar");
+                    });
+                },
+                async context =>
+                {
+                    await context.ChallengeAsync(FacebookDefaults.AuthenticationScheme);
+                    return true;
+                });
+
+            var transaction = await server.SendAsync("http://example.com/challenge");
+            var res = transaction.Response;
+
+            Assert.Equal(HttpStatusCode.Redirect, res.StatusCode);
+            Assert.Contains("scope=foo,bar", res.Headers.Location.Query);
+        }
+
+        [Fact]
+        public async Task ChallengeWillIncludeScopeAsOverwritten()
+        {
+            var server = CreateServer(
+                app => app.UseAuthentication(),
+                services =>
+                {
+                    services.AddAuthentication().AddFacebook(o =>
+                    {
+                        o.AppId = "Test App Id";
+                        o.AppSecret = "Test App Secret";
+                        o.Scope.Clear();
+                        o.Scope.Add("foo");
+                        o.Scope.Add("bar");
+                    });
+                },
+                async context =>
+                {
+                    var properties = new AuthenticationProperties();
+                    properties.Items[FacebookDefaults.ScopeParameterPropertiesKey] = "baz,qux";
```

Comment:
Don't love propagating more of this pattern everywhere... 

maybe go with a builder pattern here instead so you get intellense with the supported properties?

```
var builder = FacebookHelpers.CreateChallengePropertiesBuilder().ScopeParameterProperties = "baz,qux";
await context.ChallengeAsync("Facebook", builder.Build());
```

Reply:
Or a `FacebookProperties` that inherits from `AuthenticationProperties`?

```csharp
await context.ChallengeAsync("Facebook", new FacebookProperties
{
    Scopes = { "scope1", "scope2" }
});
```

Reply:
Yeah that's better, Builders are mostly for immuteable, so yeah just subtype for the extension methods to hang off of is good, so `FacebookChallengeProperties : AuthenticationProperties` these are purely limited to challenge right?

Reply:
If you have a sub type then they don't need to be extension methods, they can be concrete properties and methods.

Reply:
That said, do we want an explosion of sub types?

Reply:
Hmm, wouldn’t mind going in such a direction. This would actually also allow us to get back some type safety there (similar to the [proposed extension methods above](https://github.com/aspnet/Security/pull/1691#issuecomment-373643538)).

Although that would mean two new types for every authentication provider. What about simply newing-up a builder like this:

```
var builder = new FacebookChallengePropertiesBuilder()
builder.SetScope("baz", "qux");
await context.ChallengeAsync("Facebook", builder.Build());
```

And then we’d have four of those:

```
public class OAuthChallengePropertiesBuilder
{
    public ICollection<string> Scope { get; }
    public void SetScope(params string[] scopes);
    public virtual AuthenticationProperties Build();
}
public class FacebookChallengePropertiesBuilder : OAuthChallengePropertiesBuilder
{ }
public class GoogleChallengePropertiesBuilder : OAuthChallengePropertiesBuilder
{
    public string AccessTypeParameterPropertiesKey { get; set; }
    public string ApprovalPromptParameterPropertiesKey { get; set; }
    public bool IncludeGrantedScopesParameterPropertiesKey { get; set; }
    public string LoginHintParameterPropertiesKey { get; set; }
    public string Prompt { get; set; }
}
public class OpenIdConnectChallengePropertiesBuilder : OAuthChallengePropertiesBuilder
{
    public TimeSpan? MaxAge { get; set; }
    public string Prompt { get; set; }
}
```

(Microsoft Account would just use the `OAuthChallengePropertiesBuilder`)

Reply:
Yeah well the way i think of these properties bags, is that they are all just 'extension methods' on top of the Items Dictionary anyways :)

Reply:
So you really don't want to expose direct `Items[Nastyiness]` ever to the developer...

Reply:
@poke scrap the builder, its simpler just to create derived Properties... so

```
public class OAuthChallengeProperties : AuthenticationProperties
{
    public ICollection<string> Scope { get; }
    public void SetScope(params string[] scopes);
    public virtual AuthenticationProperties Build();
}
public class FacebookChallengeProperties : OAuthChallengeProperties
{ }
public class GoogleChallengeProperties : OAuthChallengeProperties
{
    public string AccessTypeParameterPropertiesKey { get; set; }
    public string ApprovalPromptParameterPropertiesKey { get; set; }
    public bool IncludeGrantedScopesParameterPropertiesKey { get; set; }
    public string LoginHintParameterPropertiesKey { get; set; }
    public string Prompt { get; set; }
}
public class OpenIdConnectChallengeProperties : OAuthChallengeProperties
{
    public TimeSpan? MaxAge { get; set; }
    public string Prompt { get; set; }
}
```

Reply:
```
var props = new FacebookChallengeProperties()
props.SetScope("baz", "qux");
await context.ChallengeAsync("Facebook", props);
```

Reply:
Hah, good point! I like that!

Reply:
@HaoK The only drawback is that for supporting getters and setters for non-string types, they will have to perform a type conversion on every call. Simple enough for the `bool` and the `TimeSpan?` but do you have a suggestion for the scope collection?

I’m currently thinking about a custom `ICollection<string>` implementation that will update the Items value properly. But maybe just having `Set~` methods will be good enough?

Reply:
Yeah you don't really need high quality getters (only ones for whatever you need to consume these on the handler side)

Reply:
I would actually keep the handler side the same as it is, to make them not dependent on those utility subclasses. There’s nothing wrong with people not using those utility classes.